### PR TITLE
feat(ivy-framework): added the `softshrink` to the ivy experimental API

### LINKED ISSUE #26095: softshrink

### PR DESCRIPTION
**Pull request by samthakur587**
# added the softshrink to the ivy experimental api.

## implemented the softshrink activation to the ivy experimental api with all the backend jax , torch , tensorflow , numpy , paddle. passing all the test. here the result.

image

Close #26095

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you run your tests and are your tests passing?
- [x] Did pre-commit not fail on any check?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by samthakur587** [comment]
hey @KareemMAX it's passing all the test in local and CI. can you please merge this so i can move to the next task. 🚀 

thanks 
sam

File: ivy/data_classes/container/experimental/activations.py [hunk 1]
```diff
@@ -1187,3 +1187,115 @@ def tanhshrink(
...
+            will be skipped. Default is ``False``.
+        prune_unapplied : bool or sequence of bools, optional
+            Whether to prune key_chains for which the function was not applied.
+            Default is ``True``.
+        map_sequences : bool or sequence of bools, optional
+            Whether to also map method to sequences (lists, tuples).
+            Default is ``False``.
+        out : array, optional
+            optional output container, for writing the result to. It must have a shape
+            that the inputs broadcast to.
```
Comment (KareemMAX) [inline ivy/data_classes/container/experimental/activations.py:1274]: Please remove all type hints, they are being added automatically by our doc builder

Comment (KareemMAX) [reply ivy/data_classes/container/experimental/activations.py:1274]: You also may remove defaults

File: ivy/functional/ivy/experimental/activations.py [hunk 2]
```diff
@@ -634,3 +634,59 @@ def tanhshrink(
...
+) -> ivy.Array:
+    """
+    Apply the softshrink function element-wise.
+
+    Parameters
+    ----------
+    x
+        input array.
+    lambd
+        the value of the lower bound of the linear region range. Default: 0.5
```
Comment (KareemMAX) [inline ivy/functional/ivy/experimental/activations.py:661]: You may remove the default from here

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py [hunk 3]
```diff
@@ -226,6 +226,34 @@ def test_silu(*, dtype_and_x, test_flags, backend_fw, fn_name, on_device):
...
 
 
+# softshrink
+@handle_test(
+    fn_tree="functional.ivy.experimental.softshrink",
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        large_abs_safety_factor=8,
+        small_abs_safety_factor=8,
+        safety_factor_scale="log",
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py:236]: Can you use `valid` instead? And add supported dtypes to the relevant functions

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py [hunk 4]
```diff
@@ -226,6 +226,34 @@ def test_silu(*, dtype_and_x, test_flags, backend_fw, fn_name, on_device):
...
+    ),
+    threshold=st.one_of(
+        st.floats(min_value=0.0, max_value=10.0),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py:239]: Why would you add limits?

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @samthakur587, everything LGTM. I've pointed out some issues, after you address them we can make the final tests and merge

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py (hunk 4 above)
Comment (samthakur587) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py:239]: In the original implementation of torch.nn.softshrink. it's takes the values of lambd > 0.0. so i put it to min value =0.0. and if you suggest then i just removed the max value. bcz it don't have the constrain on the max value.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py (hunk 3 above)
Comment (samthakur587) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py:236]: i made the changes

File: ivy/data_classes/container/experimental/activations.py (hunk 1 above)
Comment (samthakur587) [reply ivy/data_classes/container/experimental/activations.py:1274]: mode the all changes you suggested.

**Comment by samthakur587** [comment]
hi! @KareemMAX I made all the changes that you suggested. and it's passing all the test cases in local for max 150 examples. hope now it's ready to merge. 🚀 

thnaks

File: docs/demos
Comment (KareemMAX) [inline docs/demos:1]: Can you revert this change? We can't merge with it

File: ivy/data_classes/container/experimental/activations.py
Comment (KareemMAX) [inline ivy/data_classes/container/experimental/activations.py:1]: You shouldn't remove the mentions of `key_chains`, `to_apply`, etc.. Can you bring them back?

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @samthakur587,
Awesome work, just mentioned a few nitpicks and we are good to merge. Tests are passing too.

File: docs/demos
Comment (samthakur587) [reply docs/demos:1]: done 👍

File: ivy/data_classes/container/experimental/activations.py
Comment (samthakur587) [reply ivy/data_classes/container/experimental/activations.py:1]: hey @KareemMAX sorry i havn't look at this it's removed. now i added all of this .

File: ivy/data_classes/container/experimental/activations.py [hunk 5]
```diff
@@ -1188,6 +1188,125 @@ def tanhshrink(
...
+        to_apply :
+            If True, the method will be applied to key_chains, otherwise key_chains
+            will be skipped.
+        prune_unapplied :
+            Whether to prune key_chains for which the function was not applied.
+        map_sequences :
+            Whether to also map method to sequences (lists, tuples).
+        out :
+            optional output container, for writing the result to. It must have a shape
+            that the inputs broadcast to.
```
Comment (KareemMAX) [inline ivy/data_classes/container/experimental/activations.py:1281]:
```suggestion
        x
            Input container.
        lambd
            Lambda value for soft shrinkage calculation.
        key_chains
            The key-chains to apply or not apply the method to.
        to_apply
            If True, the method will be applied to key_chains, otherwise key_chains
            will be skipped.
        prune_unapplied
            Whether to prune key_chains for which the function was not applied.
        map_sequences
            Whether to also map method to sequences (lists, tuples).
        out
            optional output container, for writing the result to. It must have a shape
            that the inputs broadcast to.
```

**Review by KareemMAX — APPROVED** [review]
LGTM! just a small issue that I'll fix myself and we are good to go

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py (hunk 4 above)
Comment (KareemMAX) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py]: Oh, I just noticed that I didn't check on that issue. What would happen if you removed `max_value`? Can you experiment on that locally?

I would be more comfortable if you removed the constraint, but if it gives an error you may put it as `1e15 or something large that doesn't give an overflow error

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Sorry @samthakur587, didn't noticed the [unresolved issue](https://github.com/unifyai/ivy/pull/26493#discussion_r1349503730). Can you take a look into it?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py (hunk 4 above)
Comment (samthakur587) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_activations.py]: I tested locally it's fine upto 1e30 max_value for 125 max test examples. i added this in test to max_value = 1e30

**Comment by samthakur587** [comment]
hey @KareemMAX resolved the test.

**Review by KareemMAX — APPROVED** [review]
LGTM

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @samthakur587, tests are not passing for some reason now. I'm suspecting d07ab9a00ad35678fc16a9f8e494babeb0dae7c9 to be the issue, can you revert it?

**Comment by samthakur587** [comment]
Hi! @KareemMAX  i don't know how but the last commit i pushed from my local but with the @RashulChutani remote 😄  . I setup the new environment. and in the clone_mapping.sh file the @RashulChutani detail is there. i think my remote origin is set to his github 😄  .

**Review by KareemMAX — APPROVED** [review]
LGTM
